### PR TITLE
iDeep: F.convolution_2d and F.deconvolution_2d

### PR DESCRIPTION
**Pull request by niboshi**
iDeep support in F.convolution_2d and F.deconvolution_2d (Separated from #3865. Based on #4276.)

### CODE REVIEW

**Comment by niboshi** [comment]
Rebased.

**Comment by kmaehashi** [comment]
Please check Jenkins doctest failure. (seems like a iDeep bug?)

**Comment by niboshi** [comment]
This code causes an error.
It seems it's caused when the stride is different for x and y axes.

```
from chainer import functions as F
import numpy as np

n = 10
c_i, c_o = 3, 1
h_i, w_i = 30, 40
h_k, w_k = 10, 10
h_p, w_p = 5, 5
x = np.random.uniform(0, 1, (n, c_i, h_i, w_i)).astype('f')
W = np.random.uniform(0, 1, (c_o, c_i, h_k, w_k)).astype('f')
b = np.random.uniform(0, 1, (c_o,)).astype('f')
print(x.shape, W.shape, b.shape)

s_y, s_x = 5, 7
print("stride: ", (s_y, s_x))
print("pad: ", (h_p, w_p))

y = F.convolution_2d(x, W, b, stride=(s_y, s_x), pad=(h_p, w_p))
```
```
(10, 3, 30, 40) (1, 3, 10, 10) (1,)
stride:  (5, 7)
pad:  (5, 5)
terminate called after throwing an instance of 'mkldnn::error'
  what():  std::exception
Aborted (core dumped)
```

**Comment by niboshi** [comment]
Backtrace from gdb:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff780ef5d in __GI_abort () at abort.c:90
#2  0x00007fffd4a7db39 in __gnu_cxx::__verbose_terminate_handler () at /home/msarahan/miniconda2/conda-bld/compilers_linux-64_1507259624353/work/.build/src/gcc-7.2.0/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007fffd4a7c1fb in __cxxabiv1::__terminate (handler=<optimized out>)
    at /home/msarahan/miniconda2/conda-bld/compilers_linux-64_1507259624353/work/.build/src/gcc-7.2.0/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007fffd4a7c234 in std::terminate () at /home/msarahan/miniconda2/conda-bld/compilers_linux-64_1507259624353/work/.build/src/gcc-7.2.0/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007fffd4a7c416 in __cxxabiv1::__cxa_throw (obj=0x555556781680, tinfo=0x7fffde7c3950 <typeinfo for mkldnn::error>, dest=0x7fffde4b5ad0 <mkldnn::error::~error()>)
    at /home/msarahan/miniconda2/conda-bld/compilers_linux-64_1507259624353/work/.build/src/gcc-7.2.0/libstdc++-v3/libsupc++/eh_throw.cc:93
#6  0x00007fffde4f9740 in mkldnn::error::wrap_c_api (status=<optimized out>, message=..., error_primitive=0x0) at external/include/mkldnn.hpp:192
#7  0x00007fffde4feca2 in mkldnn::convolution_forward::desc::desc (apadding_kind=mkldnn::zero, padding_r=..., padding_l=..., dilates=..., strides=..., dst_desc=..., bias_desc=..., weights_desc=..., 
    src_desc=..., aalgorithm=mkldnn::convolution_direct, aprop_kind=mkldnn::forward_training, this=0x555556ad6850) at external/include/mkldnn.hpp:1137
#8  Convolution2DFwd<float>::setup (this=this@entry=0x555556205690, src_d=..., w_d=..., b_d=..., dst_d=..., dilate_y=dilate_y@entry=0, dilate_x=dilate_x@entry=0, s1=s1@entry=7, s2=s2@entry=5, pl1=pl1@entry=5, 
    pl2=pl2@entry=5, pr1=pr1@entry=5, pr2=pr2@entry=0) at ideep4py/primitives/ops/conv_fwd.cc:100
#9  0x00007fffde501547 in Convolution2DFwd<float>::Convolution2DFwd (this=0x555556205690, src_d=..., w_d=..., b_d=..., dst_d=..., dilate_y=0, dilate_x=0, sy=7, sx=5, pad_lh=5, pad_lw=5, pad_rh=5, pad_rw=0)
    at ideep4py/primitives/ops/conv_fwd.cc:46
#10 0x00007fffde516c86 in Convolution2DFwdFactory<float>::get (x=..., w=..., b=..., y=..., dilate_y=dilate_y@entry=0, dilate_x=dilate_x@entry=0, sy=sy@entry=7, sx=sx@entry=5, pad_lh=pad_lh@entry=5, 
    pad_lw=pad_lw@entry=5, pad_rh=pad_rh@entry=5, pad_rw=pad_rw@entry=0) at ideep4py/include/primitives/prim_mgr/conv_fwd_factory.h:56
#11 0x00007fffde517a64 in Convolution2D<float>::Forward (src=0x5555569a1410, weights=0x555556a67ec0, bias=0x555556ad4c10, cp=<optimized out>) at ideep4py/primitives/conv.cc:79
#12 0x00007fffde4a711c in Convolution2D_Py<float>::Forward (cp=<optimized out>, bias=0x555556a51090, weights=0x555556a458d0, src=0x555556a45890) at ideep4py/py/primitives/conv_py.h:55
#13 _wrap_convolution2D_Forward (self=0x0, args=<optimized out>) at ideep4py/py/ideep4py_wrap.cpp:13595
#14 0x00005555556602a1 in _PyCFunction_FastCallDict ()
#15 0x00005555556f2ebc in call_function ()
#16 0x000055555571462a in _PyEval_EvalFrameDefault ()
#17 0x00005555556ecb8b in fast_function ()
#18 0x00005555556f2f95 in call_function ()
#19 0x000055555571462a in _PyEval_EvalFrameDefault ()
#20 0x00005555556ecb8b in fast_function ()
#21 0x00005555556f2f95 in call_function ()
#22 0x000055555571462a in _PyEval_EvalFrameDefault ()
#23 0x00005555556ecb8b in fast_function ()
#24 0x00005555556f2f95 in call_function ()
#25 0x000055555571462a in _PyEval_EvalFrameDefault ()
#26 0x00005555556ec28e in _PyEval_EvalCodeWithName ()
#27 0x00005555556ecdc1 in fast_function ()
#28 0x00005555556f2f95 in call_function ()
#29 0x000055555571462a in _PyEval_EvalFrameDefault ()
#30 0x00005555556ebf24 in _PyEval_EvalCodeWithName ()
#31 0x00005555556ecdc1 in fast_function ()
#32 0x00005555556f2f95 in call_function ()
#33 0x00005555557153e7 in _PyEval_EvalFrameDefault ()
#34 0x00005555556ed8d9 in PyEval_EvalCodeEx ()
#35 0x00005555556ee67c in PyEval_EvalCode ()
#36 0x0000555555768ce4 in run_mod ()
#37 0x00005555557690e1 in PyRun_FileExFlags ()
#38 0x00005555557692e4 in PyRun_SimpleFileExFlags ()
#39 0x000055555576cdaf in Py_Main ()
#40 0x00005555556338be in main ()
```

**Comment by niboshi** [comment]
Ah, I found a bug in my implementation.

**Comment by niboshi** [comment]
`test_convolution_2d` only includes very limited cases.
I think we should improve it. (I think that can be another issue)

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -139,7 +149,34 @@ def _forward_cpu_core(self, x, W, b):
...
+        self.retain_inputs((0, 1))
+        if len(inputs) == 3:
+            x, W, b = inputs
+        else:
+            (x, W), b = inputs, None
+        out_c, input_c, kh, kw = W.shape
+        n, c, h, w = x.shape
+
+        out_h, out_w = self._get_out_size(inputs)
+        self.pd = (self.sy * (out_h - 1)
```
Comment (kmaehashi) [inline chainer/functions/connection/convolution_2d.py]: `pd` and `pr` need not to be an attribute.

File: chainer/functions/connection/convolution_2d.py [hunk 2]
```diff
@@ -318,8 +353,14 @@ def __init__(self, conv2d):
         self.cover_all = conv2d.cover_all
         self.W_dtype = W_node.dtype
         self.group = conv2d.group
+        self._use_ideep = conv2d._use_ideep
+        assert self.group == 1 or not self._use_ideep
 
     def forward_cpu(self, inputs):
+        if self._use_ideep:
+            # TODO(iDeep): Support group
```
Comment (kmaehashi) [inline chainer/functions/connection/convolution_2d.py]: Remove TODO for iDeep.

File: chainer/functions/connection/convolution_2d.py [hunk 3]
```diff
@@ -318,8 +353,14 @@ def __init__(self, conv2d):
...
+            return self.forward_ideep(inputs)
```
Comment (kmaehashi) [inline chainer/functions/connection/convolution_2d.py]: ditto

File: chainer/functions/connection/deconvolution_2d.py [hunk 4]
```diff
@@ -139,10 +141,20 @@ def forward_cpu(self, inputs):
...
-            y = self._forward_grouped_convolution(x, W, b)
+            # Grouped convolution implementation
+            return self._forward_grouped_convolution(x, W, b)
+
+        elif ((self.dy == 1 and self.dx == 1)
+              and intel64.should_use_ideep('>=auto')
+              and intel64.inputs_all_ready(inputs)):
+
+            # iDeep implementation
+            # TODO(iDeep): Support group
```
Comment (kmaehashi) [inline chainer/functions/connection/deconvolution_2d.py]: ditto

File: chainer/functions/connection/deconvolution_2d.py [hunk 5]
```diff
@@ -153,7 +165,33 @@ def _forward_cpu_core(self, x, W, b):
...
         if b is not None:
             y += b.reshape(1, b.size, 1, 1)
-        return y
+        return y,
+
+    def _forward_ideep(self, x, W, b):
+        _, in_c, kh, kw = W.shape
+        n, _, in_h, in_w = x.shape
+
+        self.pd = (self.sy * (in_h - 1)
```
Comment (kmaehashi) [inline chainer/functions/connection/deconvolution_2d.py]: ditto

**Comment by niboshi** [comment]
Thank you. Fixed for all the review comments.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**